reVX/handlers/sc_points.py

```diff
@@ -0,0 +1,631 @@
+        return self._cf_means
+
+    @staticmethod
+    def _parse_list(list_in, dtype=int):
+        """
+        Parse json list if needed
+
+        Parameters
+        ----------
+        list_in : str | list
+            List or jsonified list from supply curve table
+        dtype : np.dtype
+            dtype for output array
+
+        Returns
+        -------
+        out : ndarary
+            vector of values
+        """
+        if isinstance(list_in, str):
+            list_in = json.loads(list_in)
+        elif not isinstance(list_in, list):
+            msg = ('Must supply either a jsonified list or a list, not {}'
+                   .format(type(list_in)))
+            logger.error(msg)
+            raise ValueError(msg)
+
+        out = np.array(list_in, dtype=dtype)
+
+        return out
+
+    def _drop_build_capacity(self, build_capacity, drop=None):
+        """
+        Drop capacity from Supply Curve point in order of best resource gids
+        (based on cf_mean)
+
+        Parameters
+        ----------
+        build_capacity : float
+            Capacity to be built
+        drop : int | None, optional
+            Number of gids to drop, if None drop all, by default None
+
+        Returns
+        -------
+        sc_point : pd.Series
+            Resource gids being allocated
+        capacity : float
+            Capacity being allocated
+        availability : bool
+            Whether Supply Curve point still has available capacity
+        """
+        s = self.available_res_gids[0]
+        if drop is not None:
+            e = s + drop
+        else:
+            e = drop
+
+        drop_slice = slice(s, e, None)
+        capacity = self.capacity
+        if capacity < build_capacity:
+            build_capacity = capacity
+
+        sc_point = {'sc_gid': self.sc_gid,
+                    'res_gids': self.resource_gids[drop_slice].tolist(),
+                    'gid_counts': self.gid_counts[drop_slice].tolist(),
+                    'cf_means': self.cf_means[drop_slice].tolist(),
+                    'build_capacity': build_capacity}
+        sc_point = pd.Series(sc_point)
+
+        self._avail_cap[drop_slice] = 0.0
+        availability = self.capacity > 0
+
+        return sc_point, capacity, availability
+
+    def extract_capacity(self, capacity):
+        """
+        Extract capacity from Supply Curve point in order of best resource gids
+        (based on cf_mean)
+
+        Parameters
+        ----------
+        capacity : float
+            Capacity to extract from Supply Curve point
+
+        Returns
+        -------
+        out : tuple
+            (sc_point, capacity, availability)
+        """
+        build_capacity = capacity.copy()
+        if self.capacity > 0:
+            if capacity < self.capacity:
+                drop = 0
+                for cap in self.available_capacities:
+                    drop += 1
+                    capacity -= cap
+                    if capacity <= 0:
```

Comment:
Do you always get even builds from this? Like can it build part of a resource gid? I would think the build from each resource gid would be boolean right? This is also related to this line 212: `self._avail_cap[drop_slice] = 0.0`

Reply:
While I allow the splitting of a Supply Curve point I only let each Resource gid to be associated with a single "Plant". The logic is that we extract capacity from the resource gids until we have (over)filled the desired capacity. This is all a bit redundant as in the end we multiply the resource CF profiles by the actual desired build capacity and not the actual capacity available at each resource gid... 

For line 212, this is how I'm tracking which resource gids have been "allocated" to plants. It replaces the old np.delete logic.

Reply:
Yeah but say you have three resource gids with capacities [1, 1, 8] and you're building a plant with capacity 3. Would the final profile be weighted 80% to the third resource gid even though all three should be weighted equally? Or does that get handled somewhere else?

Reply:
That will get handled by the Aggregation class. Here is the high level logic:
When sc_gids are "allocated" to a plant their "capacities" are recomputed to sum to the desired plant build capacity.
E.g.
Plant 1 needs 100MW which will be allocated from sc_gid 1 and 2 (in that order)
sc_gid 1 has a capacity of 50MW -> all 50MW are "built"
sc_gid 2 has a capacity of 75MW -> only 50MW will be "built" as plant 1 only needs 50MW of capacity from sc_gid 2

During aggregation the resource gids capacities are computed from the gid_counts and the sc_gid build capacity.
e.g. lets say sc_gid 2 has 2 res_gids with capacities [25, 50] (recall the actual capacity of sc_gid 2 was 75MW).  But only 50MW is being built, so the res_capacities are recomputed as [17, 33].  So the distribution of capacities is retained.  

I realize that in both cases this isn't "perfect" as the largest res_gid is the last one, but on the fly changing the gid_counts or tracking the res_capacity seemed like overkill. Maybe a potential improvement down the road...